secret-rotator/rotator/rotator.go

```diff
@@ -0,0 +1,172 @@
+		return false, err
+	}
+
+	if rotatedSecret.Refresh.Interval != 0 {
+		// the refresh stratetgy is refreshInterval
+		// check the elapsed time from its createTime to now.
+		if now.After(createTime.Add(rotatedSecret.Refresh.Interval)) {
+			// provision a new secret
+			newId, newSecret, err := r.Provisioners[rotatedSecret.Type.Type()].CreateNew(labels)
+			if err != nil {
+				return false, err
+			}
+
+			// update the secret Manager secret
+			latestVersion, err := r.Client.UpsertSecret(rotatedSecret.Project, rotatedSecret.Secret, newSecret)
+			if err != nil {
+				return false, err
+			}
+
+			err = r.Client.UpsertSecretLabel(rotatedSecret.Project, rotatedSecret.Secret, "v"+latestVersion, newId)
+			if err != nil {
+				return false, err
+			}
+
+			return true, nil
+		}
+	} else {
+		// the refresh strategy is cron
+		// TODO
+	}
+
+	return false, nil
+}
+
+// Deactivate checks the old secret versions that are recorded in the secret labels,
+// if any needs to be deactivated according to 'now' and 'rotatedSecret.GracePeriod',
+// Deactivate deactivates that old version and updates the Secret Manager secret.
+func (r *SecretRotator) Deactivate(rotatedSecret config.RotatedSecretSpec, now time.Time) error {
+	labels, err := r.Client.GetSecretLabels(rotatedSecret.Project, rotatedSecret.Secret)
+	if err != nil {
+		return err
+	}
+
+	for key, _ := range labels {
+		// the key of a label for [version: data] pair should be in the format of "v%d"
+		matched, err := regexp.Match(`^v[0-9]+$`, []byte(key))
+		if err != nil {
+			return err
+		}
+
+		if !matched {
+			continue
+		}
+
+		// for each version, check the elapsed time from its next version's createTime to now.
+		curVersion := key[1:]
+		v, _ := strconv.Atoi(curVersion)
+		nextVersion := strconv.Itoa(v + 1)
+
+		// check if curVersion exists
+		err = r.Client.ValidateSecretVersion(rotatedSecret.Project, rotatedSecret.Secret, curVersion)
+		if err != nil {
+			return err
+		}
+
+		// check if nextVersion exists
+		err = r.Client.ValidateSecretVersion(rotatedSecret.Project, rotatedSecret.Secret, nextVersion)
+		if err != nil {
+			// if nextVersion does not exist, then curVersion is the latest. Do not return err.
+			if status.Code(err) == codes.NotFound {
+				continue
+			} else {
+				return err
+			}
+		}
+
+		nextCreateTime, err := r.Client.GetCreateTime(rotatedSecret.Project, rotatedSecret.Secret, nextVersion)
+		if err != nil {
+			return err
+		}
+
+		if now.After(nextCreateTime.Add(rotatedSecret.GracePeriod)) {
+			// deactivate the old secret
+			err := r.Provisioners[rotatedSecret.Type.Type()].Deactivate(labels, curVersion)
+			if err != nil {
+				return err
+			}
+
+			// TODO: disable or destroy Secret Manager secret version curVersion?
+			// 1: SecretVersion_ENABLED, 2: SecretVersion_DISABLED, 3: SecretVersion_DESTROYED
+			err = r.Client.ChangeSecretVersionState(rotatedSecret.Project, rotatedSecret.Secret, curVersion, 2)
+			if err != nil {
+				return err
+			}
```

Comment:
You could also consider a graceful deactivation path:
- first time through, active versions are disabled
- next time through, disabled versions are destroyed (and the corresponding secret deactivated)

This allows a nice grace period if there's human error involved in generating the new secret, giving us the opportunity to go back to using the old secret.

If you don't want to allow for such an option, then I don't think there's much point in keeping a secret version around as disabled if it's been deactivated.

Reply:
After discussing, if the service account key itself is deleted, there's no meaning to keep the copy of that private key.
We'll just destroy a version when it went out of its grace period.